src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -43,57 +43,65 @@ private void fetchMessages(final MessageCacheItem cacheItem, Iterator<DataSource
-    	long timestampNew = cacheItem.getTimestamp();
+    	long timestamp = cacheItem.getTimestamp();
+    	if (timestampOld == timestamp) {
+    		logger.debug(FormatUtils.format(ConstantsMsg.GET_MESSAGES_FAILED, dto.getComponent(), dto.getLocale(), dataSource.toString()));
+    	}
     	
-    	// If failed to get messages from the data source
-    	if (timestampNew == timestampOld) {
+    	// Skip this block if timestamp is not 0 (which means cacheItem is in the cache) regardless if cacheItem is expired or not.
+    	// Otherwise, try the next dataSource in the queue.
+    	if (timestamp == 0) {
     		// Try the next dataSource in the queue
     		if (msgSourceQueueIter.hasNext()) {
     			fetchMessages(cacheItem, msgSourceQueueIter);
     		// If no more data source in queue, log the error. This means that neither online nor offline fetch succeeded.
     		} else {
-    			logger.error(FormatUtils.format(ConstantsMsg.GET_MESSAGES_FAILED, dto.getComponent(), dto.getLocale(), dataSource.toString()));
+    			logger.debug(FormatUtils.format(ConstantsMsg.GET_MESSAGES_FAILED_ALL, dto.getComponent(), dto.getLocale()));
     		}
     	}
     }
     
     /**
      * Get messages from cache
      */
-    public MessageCacheItem getMessages() {
+    public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
     	CacheService cacheService = new CacheService(dto);
-    	Map<String, String> cacheOfComponent = null;
     	MessageCacheItem cacheItem = null;
     	if (cacheService.isContainComponent()) { // Item is in cache
     		cacheItem = cacheService.getCacheOfComponent();
-    		cacheOfComponent = cacheItem.getCachedData();
     		if (cacheItem.isExpired()) { // cacheItem has expired
     			// Update the cache in a separate thread
-    			populateCacheTask(cacheItem); 		
+    			populateCacheTask(cacheItem);
     		}
     	} else { // Item is not in cache
     		// Create a new cacheItem object to be stored in cache
-    		cacheItem = new MessageCacheItem();  
+    		cacheItem = new MessageCacheItem();
     		fetchMessages(cacheItem, VIPCfg.getInstance().getMsgOriginsQueue().iterator());
-    		cacheOfComponent = cacheItem.getCachedData();
+    		long timestamp = cacheItem.getTimestamp();
     		
-    		if (cacheOfComponent != null && !cacheOfComponent.isEmpty()) {
-    			cacheService.addCacheOfComponent(cacheItem);
-    		}
-    	} 
+    		if(!dto.getLocale().equals(ConstantsKeys.SOURCE)) {
+    			// If failed to fetch message, use MessageCacheItem of the next fallback locale.
+    			if (timestamp == 0 && fallbackLocalesIter.hasNext()) {
+    				MessagesDTO fallbackLocaleDTO = new MessagesDTO(dto.getComponent(), fallbackLocalesIter.next().toLanguageTag(), dto.getProductID(), dto.getVersion());
+    				cacheItem = new ComponentService(fallbackLocaleDTO).getMessages(fallbackLocalesIter);
+    			}
+				if (!cacheItem.getCachedData().isEmpty()) {
+					cacheService.addCacheOfComponent(cacheItem);
+				}
+			}
```

Comment:
Refer to the comments xiaochao added to my PR, the fallback logic should move to a new method, thus one method is for getting exact translation(without fallback),  another is for getting translation with fallback.

Reply:
Let's merge this PR and your PR first.

Reply:
Okay will do this now in this PR